### PR TITLE
DO NOT MERGE | Cosmos 4.X | Address memory leak in Direct TCP transport client

### PR DESCRIPTION
**Pull request by David-Noble-at-work**
## Repro

```bash
mkdir azure-cosmos.results
git checkout master
profile=direct
mvn -P$profile -DargLine="-Dio.netty.leakDetectionLevel=paranoid -Dio.netty.leakDetection.targetRecords=255" verify | tee azure-cosmos.results/Verify-P$profile.output
```

### Observe

A number of leak detections that implicate the Direct TCP reference counting code:

- <TODO: sample logger output>

## Fix

This is a port from v2. See https://github.com/Azure/azure-cosmosdb-java/pull/326 for a description of the change.

## TODO

Attach evidence of issue and issue resolution. This is to enable a comparison of test results with paranoid leak detection before and after the fix.

### CODE REVIEW

File: sdk/cosmos/README.md [hunk 1]
```diff
@@ -0,0 +1,295 @@
+﻿# Java SDK for SQL API of Azure Cosmos DB
```
Comment (moderakh) [inline sdk/cosmos/README.md]: why are we adding Readme file as part of this file?

Dis this come as part of merging v3 branch to v4?

File: sdk/cosmos/README.md [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+Jar dependency binary information for maven and gradle can be found here at [maven](https://mvnrepository.com/artifact/com.microsoft.azure/azure-cosmos).
+
+For example, using maven, you can add the following dependency to your maven pom file:
+
+[//]: # ({x-version-update-start;com.microsoft.azure:azure-cosmos;current})
+```xml
+<dependency>
+  <groupId>com.microsoft.azure</groupId>
+  <artifactId>azure-cosmos</artifactId>
+  <version>3.6.0</version>
```
Comment (moderakh) [inline sdk/cosmos/README.md]: v3 version. this is v4 repo.

File: sdk/cosmos/azure-cosmos-benchmark/pom.xml [hunk 3]
```diff
@@ -14,7 +14,7 @@ Licensed under the MIT License.
 
   <groupId>com.azure</groupId>
   <artifactId>azure-cosmos-benchmark</artifactId>
-  <version>4.0.1-beta.1</version> <!-- {x-version-update;com.azure:azure-cosmos-benchmark;current} -->
+  <version>4.0.1-beta.3</version> <!-- {x-version-update;com.azure:azure-cosmos-benchmark;current} -->
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos-benchmark/pom.xml]: This is an automated process for release version bump.
also we haven't released 4.0.1-beta.2 I think.
https://mvnrepository.com/artifact/com.azure/azure-cosmos

File: sdk/cosmos/README.md (hunk 1 above)
Comment (kushagraThapar) [reply sdk/cosmos/README.md]: I think so too. There are more changes on pom.xml files. I am planning to fix it now.

**Pull request merged**